bokehjs/src/coffee/models/plots/plot_canvas.coffee

```diff
@@ -304,46 +304,94 @@ class PlotCanvasView extends Renderer.View
-
-    if rng.get('start') != range_info['start'] or rng.get('end') != range_info['end']
-      rng.have_updated_interactively = true
-      rng.set(range_info)
-      rng.get('callback')?.execute(rng)
-
-  update_range: (range_info, is_panning) ->
+  _update_ranges_together: (range_info_iter) ->
+    # Get weight needed to scale the diff of the range to honour limits
+    # This is in some cases applied as a pre-processing step to _update_single_range
+    weight = 1.0
+    for [rng, range_info] in range_info_iter
+      weight = Math.min(weight, @_get_weight_to_constrain_interval(rng, range_info))
+    # Apply shared weight to all ranges
+    if weight < 1
+      for [rng, range_info] in range_info_iter
+        range_info['start'] = weight * range_info['start'] + (1-weight) * rng.get('start')
+        range_info['end'] = weight * range_info['end'] + (1-weight) * rng.get('end')
+
+  _update_ranges_individually: (range_info_iter, is_panning, is_scrolling) ->
+
+    for [rng, range_info] in range_info_iter
+      # Is this a reversed range?
+      reversed = if rng.get('start') > rng.get('end') then true else false
+
+      # Limit range interval first. Note that for scroll events,
+      # the interval has already been limited for all ranges simultaneously
+      if not is_scrolling
+        weight = @_get_weight_to_constrain_interval(rng, range_info)
+        if weight < 1
+            range_info['start'] = weight * range_info['start'] + (1-weight) * rng.get('start')
+            range_info['end'] = weight * range_info['end'] + (1-weight) * rng.get('end')
+
+      # Prevent range from going outside limits
+      # Also ensure that range keeps the same delta when panning/scrolling
+      if rng.get('bounds')?
+        min = rng.get('bounds')[0]
+        max = rng.get('bounds')[1]
+        new_interval = Math.abs(range_info['end'] - range_info['start'])
+
+        if reversed
+          if min?
+            if min >= range_info['end']
+              range_info['end'] = min
+              if is_panning? or is_scrolling?
+                range_info['start'] = min + new_interval
+          if max?
+            if max <= range_info['start']
+              range_info['start'] = max
+              if is_panning? or is_scrolling?
+                range_info['end'] = max - new_interval
+        else
+          if min?
+            if min >= range_info['start']
+              range_info['start'] = min
+              if is_panning? or is_scrolling?
+                range_info['end'] = min + new_interval
+          if max?
+            if max <= range_info['end']
+              range_info['end'] = max
+              if is_panning? or is_scrolling?
+                range_info['start'] = max - new_interval
+
+      if rng.get('start') != range_info['start'] or rng.get('end') != range_info['end']
+        rng.have_updated_interactively = true
+        rng.set(range_info)
+        rng.get('callback')?.execute(rng)
```

Comment:
This particular function was changed to do the iteration inside the function instead of outside, adds the interval constraint, and changes how the `range_info['end']` is set when `range_info['start']` is constrained, so it works well for zoom operations. Other than that, I did not change this code.